index.js

```diff
@@ -145,82 +136,58 @@ function storeOption(longOption, optionValue, options, values) {
   }
 }
 
-const parseArgs = (config = { __proto__: null }) => {
-  const args = objectGetOwn(config, 'args') ?? getMainArgs();
-  const strict = objectGetOwn(config, 'strict') ?? true;
-  const allowPositionals = objectGetOwn(config, 'allowPositionals') ?? !strict;
-  const options = objectGetOwn(config, 'options') ?? { __proto__: null };
-
-  // Validate input configuration.
-  validateArray(args, 'args');
-  validateBoolean(strict, 'strict');
-  validateBoolean(allowPositionals, 'allowPositionals');
-  validateObject(options, 'options');
-  ArrayPrototypeForEach(
-    ObjectEntries(options),
-    ({ 0: longOption, 1: optionConfig }) => {
-      validateObject(optionConfig, `options.${longOption}`);
-
-      // type is required
-      validateUnion(objectGetOwn(optionConfig, 'type'), `options.${longOption}.type`, ['string', 'boolean']);
-
-      if (ObjectHasOwn(optionConfig, 'short')) {
-        const shortOption = optionConfig.short;
-        validateString(shortOption, `options.${longOption}.short`);
-        if (shortOption.length !== 1) {
-          throw new ERR_INVALID_ARG_VALUE(
-            `options.${longOption}.short`,
-            shortOption,
-            'must be a single character'
-          );
-        }
-      }
-
-      if (ObjectHasOwn(optionConfig, 'multiple')) {
-        validateBoolean(optionConfig.multiple, `options.${longOption}.multiple`);
-      }
-    }
-  );
-
-  const result = {
-    values: { __proto__: null },
-    positionals: []
-  };
+/**
+ * Process args and turn into identified tokens:
+ * - option (along with value, if any)
+ * - positional
+ * - option-terminator
+ *
+ * @param {string[]} args, from parseArgs({ args }) or mainArgs
+ * @param {object} options - option configs, from parseArgs({ options })
+ */
+function argsToTokens(args, options) {
+  const tokens = [];
+  let index = -1;
+  let groupCount = 0;
 
   const remainingArgs = ArrayPrototypeSlice(args);
   while (remainingArgs.length > 0) {
     const arg = ArrayPrototypeShift(remainingArgs);
     const nextArg = remainingArgs[0];
+    if (groupCount > 0)
+      groupCount--;
+    else
+      index++;
 
     // Check if `arg` is an options terminator.
     // Guideline 10 in https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html
     if (arg === '--') {
-      if (!allowPositionals && remainingArgs.length > 0) {
-        throw new ERR_PARSE_ARGS_UNEXPECTED_POSITIONAL(nextArg);
-      }
-
       // Everything after a bare '--' is considered a positional argument.
-      ArrayPrototypePushApply(
-        result.positionals,
-        remainingArgs
-      );
+      ArrayPrototypePush(tokens, { kind: 'option-terminator', index });
+      ArrayPrototypeForEach(remainingArgs, (arg) =>
+        ArrayPrototypePush(
```

Comment:
why a forEach+push instead of a pushApply of a map?

Reply:
Just more familiar tools. Any particular reason to change?

Reply:
cleaner and likely faster, i guess. not a huge deal.

Reply:
Done